### ISSUE TITLE
Segmentation fault when running "make test"

### ISSUE THREAD

**Issue by rkrzr** [mischasan/aho-corasick#30]
Hi,

I wanted to give this library a try but I'm getting an error when running `make test`:

```
x@y ~/coding/aho-corasick % make all                                                                                                  master?
all done for FORMAT=''
x@y ~/coding/aho-corasick % make test                                                                                                 master?
rm -f acism_t.pass;  ./acism_t >& acism_t.fail && mv -f acism_t.fail acism_t.pass
rules.mk:159: recipe for target 'acism_t.pass' failed
make: *** [acism_t.pass] Error 139
```

Line 159 in `rules.mk` (which itself is included in `GNUmakefile`) says:

```
%.pass          : %         ; rm -f $@; ${exec.${FORMAT}} ${*D}/${*F} >& $*.fail && mv -f $*.fail $@
```

This is unfortunately a little hard to parse, but if I understand this correctly, it means that if building `acistm_t.pass` fails, then it will move the output of that build target to `acism_t.fail` instead.

`acism_t.fail` happens to exist on my machine and it contains the following segmentation fault at the end of the file:

```
      7 --
   93:7
       t -S ->S 2989
 7532:7t b=444
        h MS 31 -> 7th ->S 444
       - --
 4941:7- b=1002
        p -S ->S 1002
59726:7-p b=8293
         o -S ->S 8293
224771:7-po b=38642
          i -S ->S 38642
476955:7-poi b=161474
           n -S ->S 161474
804050:7-poin b=135688
            t MS 30 -> 7-point ->S 135688
    1435009 nnodes
    1060025 nonleaf
   88602305 inner loop
    6253096 child loop
      31412 hash collisions
     193928 hash displacements
# state machine saved as acism.tmp
./acism_t: line 7: 23154 Segmentation fault      (core dumped) $acism/acism_x $acism/words 11550973
```

I'm running this on Ubuntu 18.04 with kernel `4.15.0-36-generic #39-Ubuntu SMP Mon Sep 24 16:19:09 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux`.
`gcc` version is `gcc (Ubuntu 7.3.0-27ubuntu1~18.04) 7.3.0`.

**Comment by mischasan** [comment]
The test writes its output to *.fail, and renames the output to *.pass
if the test succeeds.
So more or less what you said.
I won't have time to look at it until the weekend.
Do you get the error if you execute "acism_x words" (i.e. no
expected-match count)?



On 18/10/2018, Robert Kreuzer <notifications@github.com> wrote:
> Hi,
>
> I wanted to give this library a try but I'm getting an error when running
> `make test`:
>
> ```
> x@y ~/coding/aho-corasick % make all
>                                                          master?
> all done for FORMAT=''
> x@y ~/coding/aho-corasick % make test
>                                                          master?
> rm -f acism_t.pass;  ./acism_t >& acism_t.fail && mv -f acism_t.fail
> acism_t.pass
> rules.mk:159: recipe for target 'acism_t.pass' failed
> make: *** [acism_t.pass] Error 139
> ```
>
> Line 159 in `rules.mk` (which itself is included in `GNUmakefile`) says:
>
> ```
> %.pass          : %         ; rm -f $@; ${exec.${FORMAT}} ${*D}/${*F} >&
> $*.fail && mv -f $*.fail $@
> ```
>
> This is unfortunately a little hard to parse, but if I understand this
> correctly, it means that if building `acistm_t.pass` fails, then it will
> move the output of that build target to `acism_t.fail` instead.
>
> `acism_t.fail` happens to exist on my machine and it contains the following
> segmentation fault at the end of the file:
>
> ```
>       7 --
>    93:7
>        t -S ->S 2989
>  7532:7t b=444
>         h MS 31 -> 7th ->S 444
>        - --
>  4941:7- b=1002
>         p -S ->S 1002
> 59726:7-p b=8293
>          o -S ->S 8293
> 224771:7-po b=38642
>           i -S ->S 38642
> 476955:7-poi b=161474
>            n -S ->S 161474
> 804050:7-poin b=135688
>             t MS 30 -> 7-point ->S 135688
>     1435009 nnodes
>     1060025 nonleaf
>    88602305 inner loop
>     6253096 child loop
>       31412 hash collisions
>      193928 hash displacements
> # state machine saved as acism.tmp
> ./acism_t: line 7: 23154 Segmentation fault      (core dumped)
> $acism/acism_x $acism/words 11550973
> ```
>
> I'm running this on Ubuntu 18.04 with kernel `4.15.0-36-generic #39-Ubuntu
> SMP Mon Sep 24 16:19:09 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux`.
> `gcc` version is `gcc (Ubuntu 7.3.0-27ubuntu1~18.04) 7.3.0`.
>
> --
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/mischasan/aho-corasick/issues/30


-- 
Engineers think equations approximate reality.
Physicists think reality approximates the equations.
Mathematicians never make the connection.

**Comment by rkrzr** [comment]
> Do you get the error if you execute "acism_x words" (i.e. no expected-match count)?

I don't get the error when I execute `acism_x words` directly.
